### PR TITLE
Thread config through handshake states

### PR DESCRIPTION
**Pull request by djc**
As suggested in #665. This does not yet make the cache interface write only. It also makes obvious that we don't actually log keys after the initial handshake; ~~I'll file an issue for that~~ I think that's correct since the master secret is logged and other secrets are derived from it anyway?

### CODE REVIEW

File: rustls/src/client/hs.rs [hunk 1]
```diff
@@ -410,6 +412,7 @@ fn emit_client_hello_for_retry(
     });
 
     let next = ExpectServerHello {
+        config: config.clone(),
```
Comment (ctz) [inline rustls/src/client/hs.rs:415]: clone not required?

**Review by ctz — COMMENTED** [review]
I'm concerned here that we're swapping one long-lived Arc::clone for many. My understanding is that these are relatively slow (like, hundreds of times slower than an Rc-flavour reference count)  because each one involves a full memory barrier which damages total memory bandwidth.

**Comment by djc** [comment]
@briansmith any thoughts on that? I don't have a good intuition for the performance of Arc clones. Do we have any benchmarks that we could use to measure any performance impact?

**Comment by briansmith** [comment]
> I'm concerned here that we're swapping one long-lived Arc::clone for many.

I think that's a valid thing to be concerned about.

AFAICT, this PR serves multiple purposes:
1. Ensure we throw away our reference to the Config when handshaking is done, or sooner, so that we don't cause it to live for the entire life of the connection. This could be done by threading the `Arc<Config>` through the handshake state machine, without breaking it up into multiple parts. This would be the same amount of `Arc`ing as previously, i.e. it is basically free. I think we should do at least this.

2. Clarify which parts of the Config are used by which parts of the handshake, and reduce the chance that some part of the handshake state machine is depending on part of the Config it shouldn't be. This is one way to do that. Another way to do that would be to have a bunch of wrapper types `Foo(Arc<Config>)`, `Bar(Arc<Config>)` that each expose a limited subset of the `Config` (an ever-shrinking subset), where we convert one type into another at each stage of the handshake. Then there would be no extra `Arc`ing.

IMO the first problem is more important to address than the second problem. I also don't think either is particularly urgent to address.

File: rustls/src/server/tls12.rs [hunk 2]
```diff
@@ -208,11 +214,15 @@ mod client_hello {
...
-            let doing_client_auth = emit_certificate_req(&mut self.transcript, cx)?;
+            let doing_client_auth = emit_certificate_req(&self.config, &mut self.transcript, cx)?;
             emit_server_hello_done(&mut self.transcript, &mut cx.common);
 
             if doing_client_auth {
                 Ok(Box::new(ExpectCertificate {
+                    key_log: self.config.key_log.clone(),
+                    verifier: self.config.verifier.clone(),
+                    ticketer: self.config.ticketer.clone(),
+                    session_storage: self.config.session_storage.clone(),
```
Comment (briansmith) [inline rustls/src/server/tls12.rs:225]: I think to make progress here, we should just keep threading the whole `config` instead of breaking it apart. If we do that then we'll be able to eliminate all the extra `clone()`s introduced and still be able to accomplish the main goal.

**Comment by djc** [comment]
As suggested, changed this to just thread the `Arc<*Config>` through the states.

**Comment by briansmith** [comment]
Commit `client: reorder states in chronological order` seems out of scope for this. I do think that's a good change but it's orthogonal to this one, and it's hard to review.

**Comment by briansmith** [comment]
Commit "client: move TLS 1.2 save_session() into ExpectFinished impl block" also seems out of scope for this PR.

**Comment by briansmith** [comment]
Is commit "server: fix incorrect rename artifacts" fixing a problem with the first commit? if so, let's merge those two commits together. Otherwise, is it a separate fix that should be in a separate PR?

**Comment by djc** [comment]
Have scoped those other commits out for now, will bring them back as a separate PR. @ctz do you have a preference for one PR vs three?

File: rustls/src/client/hs.rs [hunk 3]
```diff
@@ -118,27 +113,29 @@ fn find_session(
 pub(super) fn start_handshake(
     dns_name: webpki::DnsName,
     extra_exts: Vec<ClientExtension>,
+    config: &Arc<ClientConfig>,
```
Comment (briansmith) [inline rustls/src/client/hs.rs:116]: Wherever we're not going to clone the `Arc` then should't we use `&ClientConfig` instead of `&Arc<ClientConfig>` to be more general and to clarify the ownership/cloning?

I would expect everything to be either `config: Arc<ClientConfig>` if moving from one state to another, or `config: &ClientConfig` otherwise.

Comment (djc) [reply rustls/src/client/hs.rs:116]: `start_handshake()` is actually the entry point to the client handshake now, and it does clone the `Arc` when calling `emit_client_hello_for_retry`.

File: rustls/src/client/tls13.rs [hunk 4]
```diff
@@ -642,9 +654,9 @@ impl hs::State for ExpectCertificateVerify {
             .split_first()
             .ok_or(Error::NoCertificatesPresented)?;
         let now = std::time::SystemTime::now();
-        let cert_verified = cx
+        let cert_verified = self
             .config
-            .get_verifier()
+            .verifier
```
Comment (briansmith) [inline rustls/src/client/tls13.rs:659]: Not sure why we're removing `get_verifier()` here. Seems like an orthogonal change.

File: rustls/src/server/hs.rs [hunk 5]
```diff
@@ -290,10 +286,11 @@ pub struct ExpectClientHello {
 
 impl ExpectClientHello {
     pub fn new(
-        server_config: &ServerConfig,
+        server_config: &Arc<ServerConfig>,
```
Comment (briansmith) [inline rustls/src/server/hs.rs:289]: I think at this point we should already be doing only moves; the cloning of `server_config` should be in the caller, so that this becomes `Arc<ServerConfig>`. That is, at the state machine layer, we're always moving, never cloning.

File: rustls/src/server/tls12.rs [hunk 6]
```diff
@@ -422,10 +439,11 @@ mod client_hello {
     }
 
     fn emit_certificate_req(
+        config: &ServerConfig,
         transcript: &mut HandshakeHash,
         cx: &mut ServerContext<'_>,
     ) -> Result<bool, Error> {
-        let client_auth = cx.config.get_verifier();
+        let client_auth = config.get_verifier();
```
Comment (briansmith) [inline rustls/src/server/tls12.rs:446]: Seems strange we got rid of the `get_verifier()` for the other side but not here. I think we should revert the other side to use `get_verifier()` like before.

File: rustls/src/server/tls12.rs [hunk 7]
```diff
@@ -640,8 +664,8 @@ impl hs::State for ExpectCertificateVerify {
             let handshake_msgs = self.transcript.take_handshake_buf();
             let certs = &self.client_cert;
 
-            cx.config
-                .get_verifier()
+            self.config
+                .verifier
```
Comment (briansmith) [inline rustls/src/server/tls12.rs:668]: Why not `get_verifier()`?

File: rustls/src/server/tls13.rs [hunk 8]
```diff
@@ -296,6 +300,7 @@ mod client_hello {
             let (mut ocsp_response, mut sct_list) =
                 (server_key.get_ocsp(), server_key.get_sct_list());
             emit_encrypted_extensions(
+                &self.config,
```
Comment (briansmith) [inline rustls/src/server/tls13.rs:303]: Should `config` parameters be at the beginning or end of the argument list? Let's be consistent.

File: rustls/src/client/hs.rs (hunk 3 above)
Comment (briansmith) [reply rustls/src/client/hs.rs:116]: After reviewing the rest, I think this should be `Arc<ClientConfig>`, i.e. cloned by the caller.

**Review by briansmith — CHANGES_REQUESTED** [review]

**Comment by briansmith** [comment]
To the extent `.get_verifier()` vs. `.verifier` doesn't matter, let's revert all those changes. If we want to do it differently, let's do that in a separate PR.

**Comment by briansmith** [comment]
Another than the nits above, I think this is a very good change. Besides the progress towards making it possible to drop the `Config` earlier, I like that you were able to eliminate some apparently-redundant confusing `config` parameters, and I like that you were able to narrow the input types of some functions so they don't even need to take a whole config.

**Comment by briansmith** [comment]
> Have scoped those other commits out for now, will bring them back as a separate PR. @ctz do you have a preference for one PR vs three?

If you do the renaming/typo fixing one as a separate PR then I'll review it ASAP. The other changes that reorder the functions are better reviewed by @ctz.

File: rustls/src/client/hs.rs (hunk 3 above)
Comment (djc) [reply rustls/src/client/hs.rs:116]: Not sure I agree. We don't need to clone it sooner, and there is value in not cloning it if we're going to error out of `start_handshake()`, for example.

File: rustls/src/client/hs.rs (hunk 1 above)
Comment (djc) [reply rustls/src/client/hs.rs:415]: Sorry, I missed that. Fixed it up now.

File: rustls/src/client/tls13.rs (hunk 4 above)
Comment (djc) [reply rustls/src/client/tls13.rs:659]: Added it back.

File: rustls/src/server/tls12.rs (hunk 6 above)
Comment (djc) [reply rustls/src/server/tls12.rs:446]: Done.

File: rustls/src/server/tls12.rs (hunk 7 above)
Comment (djc) [reply rustls/src/server/tls12.rs]: Changed it back. Will add another changes to get rid of `get_verifier()` to another PR.

File: rustls/src/server/hs.rs (hunk 5 above)
Comment (djc) [reply rustls/src/server/hs.rs]: Done.

File: rustls/src/client/hs.rs (hunk 3 above)
Comment (djc) [reply rustls/src/client/hs.rs:116]: Changed it anyway for consistency.

File: rustls/src/client/hs.rs (hunk 1 above)
Comment (briansmith) [reply rustls/src/client/hs.rs:415]: I still see the clone.

File: rustls/src/server/tls13.rs (hunk 8 above)
Comment (djc) [reply rustls/src/server/tls13.rs]: Argument order here is a bit of mess anyway, moved it to the end.

**Comment by briansmith** [comment]
@djc Not sure what's going on with the review discussion. IIRC the only outstanding issue in this PR is this:

```
let next = ExpectServerHello {
        config: config.clone(),
```
IIUC, the `clone()` isn't needed and should be removed. I tried to fix this for you in https://github.com/djc/rustls/pull/1 but maybe you didn't noticed it, and now my PR is out of date w.r.t. your branch. Or maybe I'm misunderstanding this?

**Comment by djc** [comment]
Huh, I was pretty sure I fixed that. Maybe I made a rebase error. Please check.

**Comment by briansmith** [comment]
@djc When reviewing your PR again, I noticed that we can avoid the clones completely. I was able to modify your PR so that there are no clones of the configs at all. See https://github.com/djc/rustls/pull/2 for the server. I have an analogous change for the client.

The rule of thumb that I follow is to never clone an `Arc` if it is reasonable for the caller to clone the `Arc` themselves, if avoiding the cloning has the potential to eliminate cloning completely.

**Review by briansmith — APPROVED** [review]
